### PR TITLE
Contract editor toolboxes when not in use

### PR DESCRIPTION
**Pull request by peppy**
Follow up to #16327. Reusing existing UX which seems like a good idea until we have something better. Truncation is a bit weird but I think it's fine for now.

https://user-images.githubusercontent.com/191335/148381234-0bd7ff81-e472-46fc-9165-de7d0f1a9c5e.mp4

### CODE REVIEW

**Comment by bdach** [comment]
The contracted headings don't look super great:

2022-01-06-134745_120x386_scrot

Maybe they should be hidden entirely, or show only the hotkeys in contracted mode (as in, say 1-9 / Q~P respectively), and only show the full text when expanded?

**Comment by peppy** [comment]
That's what I meant by the weird truncation. I'm not too annoyed by it, but can try hiding if required.

File: osu.Game/Overlays/ExpandingButtonContainer.cs [hunk 1]
```diff
@@ -121,15 +115,13 @@ public ExpandedState State
 
         private Drawable lastHoveredButton;
 
-        private Drawable hoveredButton => content.Children.FirstOrDefault(c => c.IsHovered);
+        private Drawable hoveredButton => FillFlow.ChildrenOfType<OsuButton>().FirstOrDefault(c => c.IsHovered);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Overlays/ExpandingButtonContainer.cs:118]: Unfortunately for the editor case, the buttons are quite deep inside the usage. Not sure there's a better way to do this.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Overlays/ExpandingButtonContainer.cs:118]: I suppose so. I guess I could look away and treat this as an exception (like the skin editor, which also uses `ChildrenOfType<>()` upon double checking).

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Overlays/ExpandingButtonContainer.cs]: I've reduced the invocations of this anyway, and added an inline comment explaining why it's there.

File: osu.Game/Overlays/SettingsToolboxGroup.cs [hunk 2]
```diff
@@ -132,6 +134,16 @@ protected SettingsToolboxGroup(string title)
         private const float fade_duration = 800;
         private const float inactive_alpha = 0.5f;
 
+        protected override bool OnInvalidate(Invalidation invalidation, InvalidationSource source)
+        {
+            // These toolbox grouped may be contracted to only show icons.
+            // For now, let's hide the header to avoid text truncation weirdness in such cases.
+            if (invalidation.HasFlagFast(Invalidation.DrawSize))
+                headerText.FadeTo(headerText.DrawWidth < DrawWidth ? 1 : 0, 150, Easing.OutQuint);
```
Comment (bdach) [inline osu.Game/Overlays/SettingsToolboxGroup.cs:142]: I'm fuzzy on the details of that invalidation subsystem again, but wasn't the point of `OnInvalidate()` that at that point layout and the properties related to it have invalid values at that point and therefore that `headerText.DrawWidth < DrawWidth` check shouldn't/can't occur there? As in `OnInvalidate()` implementations should only mark that there's a change coming, but the actual re-validation/re-layout should happen on the next `Update()` call?

Comment (smoogipoo) [reply osu.Game/Overlays/SettingsToolboxGroup.cs:142]: This is correct, `OnInvalidate` should never refer to `DrawWidth`.

**Review by bdach — APPROVED** [review]
Let's give it a go and see how users react.

**Pull request merged**